### PR TITLE
System Plugins - Update Yosemite layer - Add Action and Store - 2/2

### PR DESCRIPTION
**Pull request by fernandofgfer**
#4646 2/2

## Summary
This PR has been splitter into two PRs.  
This PR contains the Yosemite updates of the System plugins project.  
In this branch, I've included the next classes
- SystemPluginAction: To define both actions.
- SystemPluginStore: To execute these actions.

## Actions
I've added two actions:
- synchronizeSystemPlugins: This action retrieves all plugins and saves them into specific storage.
- fetchSystemPlugins: This action returns all plugins saved in storage.

## How to test it
I've added unit tests for SystemPluginStore.
Execute all unit tests

Update release notes:
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Yosemite/Yosemite/Actions/SystemPluginAction.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+///
+public enum SystemPluginAction: Action {
+
+    /// Synchronize all system plugins for a site given its ID
+    ///
+    case synchronizeSystemPlugins(siteID: Int64, onCompletion: (Result<Void, Error>) -> Void)
+
+    /// Fetch all system plugins for a site given its ID
+    ///
+    case fetchSystemPlugins(siteID: Int64, onCompletion: ([SystemPlugin]?) -> Void)
```
Comment (Ecarrion) [inline Yosemite/Yosemite/Actions/SystemPluginAction.swift]: Could you explain the use case for this action? I don't think we need it 🤔

File: Yosemite/Yosemite/Stores/SystemPluginStore.swift [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        // remove stale system plugins
+        let currentSystemPlugins = readonlySystemPlugins.map(\.name)
+        storage.deleteStaleSystemPlugins(siteID: siteID, currentSystemPlugins: currentSystemPlugins)
+    }
+
+    /// Retrieve `SystemPlugin` entities of a specified storage by siteID
+    ///
+    func fetchSystemPlugins(siteID: Int64, completionHandler: @escaping ([SystemPlugin]?) -> Void) {
+        let writerStorage = storageManager.writerDerivedStorage
+        let systemPlugins = writerStorage.loadSystemPlugins(siteID: siteID).map { $0.toReadOnly() }
```
Comment (Ecarrion) [inline Yosemite/Yosemite/Stores/SystemPluginStore.swift]: ⚠️ We should not be using a writer-derived storage to read data. It's probably better to use `storageManager.viewStorage`

**Review by Ecarrion — CHANGES_REQUESTED** [review]
Hi @fernandofgfer, This looks good! I only left a small concern to address!

Suggestion: What do you think about naming these `Actions` and `Store` something like `SystemStatusAction` and `SystemStatusStore`. The reason is to not having to create multiple new stores when in the future we need to fetch more information from that endpoint!

File: Yosemite/Yosemite/Actions/SystemPluginAction.swift (hunk 1 above)
Comment (fernandofgfer) [reply Yosemite/Yosemite/Actions/SystemPluginAction.swift]: Hi! As we talked about, in the tool that we will build to know if a feature could be enabled by his plugins, we will need something to get all plugins. So I think we will need a method to retrieve the plugins saved in the storage layer.

Comment (Ecarrion) [reply Yosemite/Yosemite/Actions/SystemPluginAction.swift]: Sure, but why we would need to fetch all of them? I think fetching just one should be enough, don't you think?

Comment (fernandofgfer) [reply Yosemite/Yosemite/Actions/SystemPluginAction.swift]: Yes, If we do the tools in that way, of course. In the end, I think it depends on if you want to load all plugins and check if a feature is enabled with that list, or if you want to search a specific plugin to know if a feature should be enabled or not. I will change with the second option.

**Comment by fernandofgfer** [comment]
> Hi @fernandofgfer, This looks good! I only left a small concern to address!
> 
> Suggestion: What do you think about naming these `Actions` and `Store` something like `SystemStatusAction` and `SystemStatusStore`. The reason is to not having to create multiple new stores when in the future we need to fetch more information from that endpoint!

Nice catch!! It's a good idea to reuse the store and the action in the future. Thanks!

**Comment by fernandofgfer** [comment]
Hi @Ecarrion. Changes added!

File: Yosemite/Yosemite/Actions/SystemStatusAction.swift [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+///
+public enum SystemStatusAction: Action {
+
+    /// Synchronize all system plugins for a site given its ID
+    ///
+    case synchronizeSystemPlugins(siteID: Int64, onCompletion: (Result<Void, Error>) -> Void)
+
+    /// Fetch an specific systemPlugin by siteID and name
+    ///
+    case fetchSystemPlugins(siteID: Int64, systemPluginName: String, onCompletion: (SystemPlugin?) -> Void)
```
Comment (Ecarrion) [inline Yosemite/Yosemite/Actions/SystemStatusAction.swift]:
```suggestion
    case fetchSystemPlugin(siteID: Int64, systemPluginName: String, onCompletion: (SystemPlugin?) -> Void)
```

**Review by Ecarrion — APPROVED** [review]
Good Job, the only thing I see missing is that this `Store` is not yet registered in our `AuthenticatedState`. And you won't be able to use it unless you register it.

**Comment by fernandofgfer** [comment]
> Good Job, the only thing I see missing is that this `Store` is not yet registered in our `AuthenticatedState`. And you won't be able to use it unless you register it.

Thanks. I will add it in the next PR!

**Review by Ecarrion — APPROVED** [review]

**Pull request merged**